{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c429a05e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gtfparse import read_gtf\n",
    "import pysam as ps\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f602fc44",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix_variant_file = 'transcripts_under_10_percent_expression'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57fe0a99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert plasmid positions into genomic positions\n",
    "\n",
    "# plasmid to genomic position table\n",
    "plasmid_gen_table = pd.read_excel('00_input_data/genom_plasmid_pos_table.xlsx')\n",
    "\n",
    "# variant table\n",
    "in_plasmid_pos = pd.read_excel('00_input_data/'+prefix_variant_file+'_in_plasmid_pos.xlsx')\n",
    "\n",
    "# define output file name\n",
    "out_pos_file = '02_output_analysis/00_converted_pos/'+prefix_variant_file+'_for_cDNA.xlsx'\n",
    "\n",
    "def convert_position(test_position, plasmid_genomic_table = plasmid_gen_table):\n",
    "    converted_pos = 0\n",
    "    \n",
    "    for strand, start_g, end_g, start_p, end_p in zip(\n",
    "        plasmid_genomic_table['strand'], plasmid_genomic_table['start_gen'], plasmid_genomic_table['end_gen'],\n",
    "        plasmid_genomic_table['start_plasmid'], plasmid_genomic_table['end_plasmid']):\n",
    "        \n",
    "        if test_position >= start_p and test_position <= end_p:\n",
    "            # distance to start position of exon/intron\n",
    "            dist_start = test_position - start_p\n",
    "            if strand == 'minus':\n",
    "                # converted position is end position of genomic exon/intron minus distance\n",
    "                converted_pos = end_g - dist_start\n",
    "            elif strand == 'plus':\n",
    "                # converted position is start position of genomic exon/intron plus distance\n",
    "                converted_pos = start_g + dist_start\n",
    "            else:\n",
    "                print('Check if strand column is filled correctly (minus or plus).')\n",
    "        \n",
    "    return converted_pos\n",
    "\n",
    "\n",
    "def output_position_file(out_positions_file, in_plasmid_positions = in_plasmid_pos):\n",
    "    # SDHB specific (if strand == 'plus': exchange pos_last and pos_next)\n",
    "    in_plasmid_positions['pos_next'] = in_plasmid_positions['plasmid_pos_last'].apply(convert_position)\n",
    "    in_plasmid_positions['pos_last'] = in_plasmid_positions['plasmid_pos_next'].apply(convert_position)\n",
    "\n",
    "    in_plasmid_positions['strand'] = 'minus'\n",
    "    in_plasmid_positions['HUGO_Symbol'] = 'SDHB'\n",
    "    in_plasmid_positions['kind'] = in_plasmid_positions['kind'].str.lower()\n",
    "\n",
    "    in_plasmid_positions[['#CHROM', 'POS', 'REF', 'ALT', 'pos_last', 'pos_next', 'kind', 'shortcut', \n",
    "                          'strand', 'HUGO_Symbol']].to_excel(out_positions_file, index=False)\n",
    "\n",
    "output_position_file(out_pos_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d395189d",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_table_name = out_pos_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fe86ed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# whole reference sequence\n",
    "ref_seq = ps.FastaFile('Hsapiens/GRCh38/ensembl/Homo_sapiens.GRCh38.dna.primary_assembly.fa')\n",
    "\n",
    "# table with variants and positions\n",
    "variants = pd.read_excel(variant_table_name)\n",
    "# which reference transcript\n",
    "variants['RefSeq accession'] = 'NM_003000'\n",
    "\n",
    "# GTF file with sequence information\n",
    "# gtf_df = read_gtf('Hsapiens/GRCh38/ensembl/GRCh38_latest_genomic.gtf')\n",
    "# only those reference transcripts within variant file\n",
    "# sdhb_gtf = gtf_df[(gtf_df['transcript_id'].str.startswith('NM_003000'))&(gtf_df['feature']=='CDS')]\n",
    "# sdhb_gtf['transcript_id'] = sdhb_gtf['transcript_id'].str.split('.', expand=True)[0]\n",
    "# sdhb_gtf['seqname'] = sdhb_gtf['seqname'].str.split('1.1', expand=True)[1]\n",
    "# sdhb_gtf.to_pickle('sdhb_gtf_pickle')\n",
    "\n",
    "# sdhb_gtf = pd.read_pickle('01_dataframes/sdhb_gtf_pickle').reset_index(drop=True)\n",
    "# sdhb_gtf['diff_end_start'] = sdhb_gtf['end'] - sdhb_gtf['start'] + 1\n",
    "# sdhb_gtf['exon_number'] = sdhb_gtf['exon_number'].astype(int)\n",
    "\n",
    "# list_comb_cod_end = []\n",
    "# for num in sdhb_gtf['exon_number']:\n",
    "#     in_diff_sum = 0\n",
    "#     for j in range(num):\n",
    "#         in_diff_sum += sdhb_gtf[sdhb_gtf['exon_number']==j+1].iloc[0]['diff_end_start']\n",
    "#     list_comb_cod_end.append(in_diff_sum)\n",
    "    \n",
    "# sdhb_gtf['end_coding_num'] = list_comb_cod_end\n",
    "# sdhb_gtf['start_coding_num'] = sdhb_gtf['end_coding_num'] - sdhb_gtf['diff_end_start'] + 1\n",
    "# sdhb_gtf[['feature','start','end','exon_number', 'diff_end_start', 'end_coding_num', 'start_coding_num']]\n",
    "\n",
    "# sdhb_gtf.to_pickle('01_dataframes/sdhb_gtf_pickle')\n",
    "\n",
    "# GTF file with sequence information for SDHB\n",
    "sdhb_gtf = pd.read_pickle('01_dataframes/sdhb_gtf_pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee30e04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_col(df, inp):\n",
    "    return [i for i in df.columns if inp in i.lower()]\n",
    "\n",
    "def rev_comp(seq):\n",
    "    return str(Seq(seq).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5c5aa5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# new cDNA/protein sequence with indels and variants\n",
    "def cDNA(ind):\n",
    "    cdna = ''\n",
    "    ind_df = sdhb_gtf\n",
    "    kind = variants.iloc[ind]['kind']\n",
    "    chrom = str(variants.iloc[ind]['#CHROM'])\n",
    "    var_pos = int(variants.iloc[ind]['POS'])\n",
    "    ref_len = len(variants.iloc[ind]['REF'])-1\n",
    "    pos_last = int(variants.iloc[ind]['pos_last'])\n",
    "    pos_next = int(variants.iloc[ind]['pos_next'])\n",
    "    ori = list(ind_df['strand'])[0]\n",
    "    aa_pos = 0\n",
    "    hgsv_pos = 0\n",
    "    \n",
    "    # exclude duplicated sequences\n",
    "    new_pos = 0\n",
    "    \n",
    "    # for new HGVSg in VCF\n",
    "    ref_n = ''\n",
    "    alt_n = ''\n",
    "    \n",
    "    # for coding positions, differences\n",
    "    cod_start_n = ''\n",
    "    cod_end_n = ''\n",
    "    cod_diff = 0\n",
    "    \n",
    "    cdna_hgvs = ''\n",
    "    \n",
    "    for num,s,e,s_c,e_c in zip(ind_df.index, ind_df['start'], ind_df['end'], ind_df['start_coding_num'], \n",
    "                               ind_df['end_coding_num']):\n",
    "        # for insertion (new pos (start/end of new exon) in intron)\n",
    "        if new_pos == 0:\n",
    "            if pos_last == e:\n",
    "                e = pos_next-1\n",
    "                new_pos +=1\n",
    "                ref_n = ref_seq.fetch(chrom, pos_last-1, pos_last)\n",
    "                alt_n = ref_seq.fetch(chrom, pos_last-1, pos_next-1)\n",
    "                hgsv_pos = pos_last\n",
    "                # include coding position and HGVS\n",
    "                if ori == '+' or ori == 'plus':\n",
    "                    continue\n",
    "                elif ori == '-' or ori == 'minus':\n",
    "                    cod_start_n = s_c - 1\n",
    "                    cod_end_n = s_c\n",
    "                    cod_diff = pos_next - pos_last - 1\n",
    "                    #cdna_hgvs = f'{cod_start_n}_{cod_end_n}ins[{cod_start_n}-{cod_diff}_{cod_start_n}-1]'\n",
    "                    cdna_hgvs = f'{cod_start_n}_{cod_end_n}ins[{cod_end_n}-{cod_diff}_{cod_end_n}-1]'\n",
    "                    \n",
    "            elif pos_next == s:\n",
    "                s = pos_last+1\n",
    "                new_pos +=1\n",
    "                ref_n = ref_seq.fetch(chrom, pos_next-2, pos_next-1)\n",
    "                alt_n = ref_n + ref_seq.fetch(chrom, pos_last, pos_next-1)\n",
    "                hgsv_pos = pos_next-1\n",
    "                # include coding position and HGVS\n",
    "                if ori == '+' or ori == 'plus':\n",
    "                    continue\n",
    "                elif ori == '-' or ori == 'minus':\n",
    "                    cod_start_n = e_c\n",
    "                    cod_end_n = e_c + 1\n",
    "                    cod_diff = pos_next - pos_last - 1\n",
    "                    cdna_hgvs = f'{cod_start_n}_{cod_end_n}ins[{cod_start_n}+1_{cod_start_n}+{cod_diff}]'\n",
    "                \n",
    "            # for deletion (new pos (start/end of new exon) in exon)\n",
    "            elif pos_next-1 == e:\n",
    "                e = pos_last\n",
    "                new_pos +=1\n",
    "                ref_n = ref_seq.fetch(chrom, pos_last-1, pos_next-1)\n",
    "                alt_n = ref_seq.fetch(chrom, pos_last-1, pos_last)\n",
    "                hgsv_pos = pos_last\n",
    "                # include coding position and HGVS\n",
    "                if ori == '+' or ori == 'plus':\n",
    "                    continue\n",
    "                elif ori == '-' or ori == 'minus':\n",
    "                    cod_diff = pos_next - pos_last - 1\n",
    "                    cod_start_n = s_c\n",
    "                    cod_end_n = s_c + cod_diff - 1\n",
    "                    cdna_hgvs = f'{cod_start_n}_{cod_end_n}del'\n",
    "                \n",
    "            elif pos_last+1 == s:\n",
    "                s = pos_next\n",
    "                new_pos +=1\n",
    "                ref_n = ref_seq.fetch(chrom, pos_last-1, pos_next-1)\n",
    "                alt_n = ref_seq.fetch(chrom, pos_last-1, pos_last)\n",
    "                hgsv_pos = pos_last\n",
    "                # include coding position and HGVS\n",
    "                if ori == '+' or ori == 'plus':\n",
    "                    continue\n",
    "                elif ori == '-' or ori == 'minus':\n",
    "                    cod_diff = pos_next - pos_last - 1\n",
    "                    cod_start_n = e_c - cod_diff + 1\n",
    "                    cod_end_n = e_c\n",
    "                    cdna_hgvs = f'{cod_start_n}_{cod_end_n}del'\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "        # e.g. in exon skipping events\n",
    "        if s+1 > e:\n",
    "            cdna += ''\n",
    "        \n",
    "        # if original variant is in new cDNA\n",
    "        elif s <= var_pos and e >= var_pos:\n",
    "            # for deletions starting in exon\n",
    "            if ref_len>0:\n",
    "                if ori == '+':\n",
    "                    if var_pos+ref_len < e:\n",
    "                        cdna += ref_seq.fetch(chrom, s-1, var_pos)\n",
    "                        cdna += ref_seq.fetch(chrom, var_pos+ref_len, e)\n",
    "                    else: # var_pos+ref_len >= e\n",
    "                        cdna += ref_seq.fetch(chrom, s-1, var_pos)\n",
    "                elif ori == '-' or ori == 'minus':\n",
    "                    if var_pos+ref_len < e:\n",
    "                        cdna += rev_comp(ref_seq.fetch(chrom, var_pos+ref_len, e))\n",
    "                        cdna += rev_comp(ref_seq.fetch(chrom, s-1, var_pos))\n",
    "                    else: # var_pos+ref_len >= e\n",
    "                        cdna += rev_comp(ref_seq.fetch(chrom, s-1, var_pos))\n",
    "                else:\n",
    "                    pass\n",
    "\n",
    "            # for insertions, substitutions\n",
    "            else:\n",
    "                if ori == '+':\n",
    "                    cdna += ref_seq.fetch(chrom, s-1, var_pos-1)\n",
    "                    cdna += variants.iloc[ind]['ALT']\n",
    "                    cdna += ref_seq.fetch(chrom, var_pos, e)\n",
    "                elif ori == '-' or ori == 'minus':\n",
    "                    cdna += rev_comp(ref_seq.fetch(chrom, var_pos, e))\n",
    "                    cdna += rev_comp(variants.iloc[ind]['ALT'])\n",
    "                    cdna += rev_comp(ref_seq.fetch(chrom, s-1, var_pos-1))\n",
    "                else:\n",
    "                    pass\n",
    "    \n",
    "        # for deletions starting in intron\n",
    "        elif ref_len>0 and s <= var_pos+ref_len and e >= var_pos+ref_len:\n",
    "            if ori == '+':\n",
    "                cdna += ref_seq.fetch(chrom, var_pos+ref_len, e)\n",
    "            elif ori == '-' or ori == 'minus':\n",
    "                cdna += rev_comp(ref_seq.fetch(chrom, var_pos+ref_len, e))\n",
    "            else:\n",
    "                pass\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        elif ori == '+':\n",
    "            cdna += ref_seq.fetch(chrom, s-1, e)\n",
    "        elif ori == '-' or ori == 'minus':\n",
    "            cdna += rev_comp(ref_seq.fetch(chrom, s-1, e))\n",
    "        else:\n",
    "            pass\n",
    "    return [cdna, hgsv_pos, ref_n, alt_n, cdna_hgvs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a755b834",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for reference cDNA/protein sequence for comparison\n",
    "def ref_cdna(ind):\n",
    "    cdna = ''\n",
    "    chrom = str(variants.iloc[ind]['#CHROM'])\n",
    "    ind_df = sdhb_gtf\n",
    "    ori = list(ind_df['strand'])[0]\n",
    "    for s,e in zip(ind_df['start'], ind_df['end']):\n",
    "        if ori == '+':\n",
    "            cdna += ref_seq.fetch(chrom, s-1, e)\n",
    "        elif ori == '-' or ori == 'minus':\n",
    "            cdna += rev_comp(ref_seq.fetch(chrom, s-1, e))\n",
    "        else:\n",
    "            pass\n",
    "    return cdna\n",
    "\n",
    "# position of change in nucleic sequence\n",
    "def nuc_pos(ind):\n",
    "    alt_nuc = variants.iloc[ind]['cDNA']\n",
    "    ref_nuc = variants.iloc[ind]['ref_cDNA']\n",
    "    len_indel = abs(len(alt_nuc)-len(ref_nuc))\n",
    "    ct = 0\n",
    "    nuc_pos = 0\n",
    "    for a,r in zip(alt_nuc, ref_nuc[:len(alt_nuc)]):\n",
    "        ct+=1\n",
    "        if a != r:\n",
    "            nuc_pos = ct\n",
    "            break\n",
    "        else:\n",
    "            continue\n",
    "    if nuc_pos == 0 and len(alt_nuc)<len(ref_nuc):\n",
    "        nuc_pos = len(alt_nuc)+1\n",
    "    elif nuc_pos == 0 and len(alt_nuc)==len(ref_nuc):\n",
    "        nuc_pos = 'no_change'\n",
    "    return [nuc_pos, len_indel]\n",
    "\n",
    "# position of change in aa sequence\n",
    "def prot_pos(ind):\n",
    "    alt_prot = variants.iloc[ind]['termin_prot']\n",
    "    ref_prot = variants.iloc[ind]['ref_protein_seq']\n",
    "    ct = 0\n",
    "    aa_pos = 0\n",
    "    for a,r in zip(alt_prot, ref_prot[:len(alt_prot)]):\n",
    "        ct+=1\n",
    "        if a != r:\n",
    "            aa_pos = ct\n",
    "            break\n",
    "        else:\n",
    "            continue\n",
    "    if aa_pos == 0 and len(alt_prot)<len(ref_prot):\n",
    "        aa_pos = len(alt_prot)+1\n",
    "    elif aa_pos == 0 and len(alt_prot)==len(ref_prot):\n",
    "        aa_pos = 'no_change'\n",
    "    return aa_pos\n",
    "\n",
    "# length of exons\n",
    "def exon_len(ind):\n",
    "    ind_df = sdhb_gtf\n",
    "    exon_lst = []\n",
    "    bott = 0\n",
    "    for s,e,n in zip(ind_df['start'], ind_df['end'], ind_df['exon_number']):\n",
    "        exon_lst.append((n, bott+1, bott+abs(s-e)+1))\n",
    "        bott += abs(s-e)+1\n",
    "    return exon_lst, n\n",
    "\n",
    "# in which exon change occurs\n",
    "def changed_exon(ind):\n",
    "    exon = 0\n",
    "    nuc_change_pos = variants.iloc[ind]['nuc_change_pos']\n",
    "    if type(nuc_change_pos) != 'no_change':\n",
    "        for ex,s,e in variants.iloc[ind]['exons_lengths']:\n",
    "            if nuc_change_pos >= s and nuc_change_pos <= e:\n",
    "                exon = ex\n",
    "            else:\n",
    "                pass\n",
    "    return exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0951b98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find out next matching amino acids position\n",
    "import re\n",
    "def aas_next(ind):\n",
    "    change_pos = variants.iloc[ind]['prot_change_pos']\n",
    "    ref_n = variants.iloc[ind]['ref_n']\n",
    "    alt_n = variants.iloc[ind]['alt_n']\n",
    "    len_indel_nuc = abs(len(ref_n)-len(alt_n))\n",
    "    rest_3 = len_indel_nuc%3\n",
    "    ct = 0\n",
    "    pos_next_normal = ''\n",
    "    len_indel = 0\n",
    "    \n",
    "    try:\n",
    "        if type(change_pos)!='no_change':\n",
    "            alt1 = variants.iloc[ind]['termin_prot']\n",
    "            short_alt = alt1[change_pos-1:]\n",
    "            ref1 = variants.iloc[ind]['ref_protein_seq']\n",
    "            short_ref = ref1[change_pos-1:]\n",
    "            kind = variants.iloc[ind]['kind']\n",
    "            \n",
    "            if kind=='ins':\n",
    "                if rest_3 == 0:\n",
    "                    if re.search(short_ref, short_alt) and len(alt1)>len(ref1):\n",
    "                        len_indel = re.search(short_ref, short_alt).regs[0][0]\n",
    "                        pos_next_normal = change_pos+len_indel\n",
    "                    elif re.search(short_ref[1:], short_alt) and len(alt1)>len(ref1):\n",
    "                        len_indel = re.search(short_ref[1:], short_alt).regs[0][0]\n",
    "                        pos_next_normal = change_pos+len_indel\n",
    "                else:\n",
    "                    pos_next_normal = 'probably frameshift'\n",
    "            \n",
    "            elif kind =='del':\n",
    "                if rest_3 == 0:\n",
    "                    if re.search(short_alt, short_ref) and len(ref1)>len(alt1):\n",
    "                        len_indel = re.search(short_alt, short_ref).regs[0][0]\n",
    "                        pos_next_normal = change_pos+len_indel\n",
    "                    elif re.search(short_alt[1:], short_ref) and len(ref1)>len(alt1):\n",
    "                        len_indel = re.search(short_alt[1:], short_ref).regs[0][0]\n",
    "                        pos_next_normal = change_pos+len_indel\n",
    "                else:\n",
    "                    pos_next_normal = 'probably frameshift'\n",
    "            else:\n",
    "                pass\n",
    "        else:\n",
    "            pos_next_normal = 'no change'\n",
    "        if abs(len(alt1)-len(ref1)) > (len_indel+1):\n",
    "            pos_next_normal = 'probably frameshift'\n",
    "            \n",
    "        else:\n",
    "            pass\n",
    "    except:\n",
    "        pos_next_normal = 'error'\n",
    "    return [pos_next_normal, len_indel]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c5661b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# new columns in variants df with cDNA, protein sequence\n",
    "variants[['cDNA', 'hgsv_pos', 'ref_n', 'alt_n', 'cDNA_HGVS']] = [cDNA(i) for i in variants.index]\n",
    "variants['protein_seq'] = [str(Seq(cDNA(i)[0]).translate()) for i in variants.index]\n",
    "# terminated protein sequence after first stop codon\n",
    "variants['termin_prot'] = variants['protein_seq'].str.split('*', expand=True)[0]\n",
    "\n",
    "variants['cDNA_HGVS'] = 'NM_003000:r.' + variants['cDNA_HGVS']\n",
    "\n",
    "# new columns in variants df with reference cDNA, protein sequence\n",
    "variants['ref_cDNA'] = [ref_cdna(i) for i in variants.index]\n",
    "variants['ref_protein_seq'] = [str(Seq(ref_cdna(i)).translate()) for i in variants.index]\n",
    "\n",
    "# length of sequences\n",
    "for i in ['cDNA', 'termin_prot', 'ref_cDNA', 'ref_protein_seq']:\n",
    "    variants[i+'_len'] = variants[i].str.len()\n",
    "\n",
    "# position of change in nucleic/aa sequence\n",
    "variants[['nuc_change_pos', 'len_indel_nuc']] = [nuc_pos(i) for i in variants.index]\n",
    "variants['prot_change_pos'] = [prot_pos(i) for i in variants.index]\n",
    "\n",
    "# position of next matching aa and length of indel\n",
    "variants[['aa_next_pos', 'len_indel_substit_prot']] = [aas_next(i) for i in variants.index]\n",
    "\n",
    "# length of exons\n",
    "variants[['exons_lengths', 'max_exon']] = [exon_len(i) for i in variants.index]\n",
    "\n",
    "# changed exon\n",
    "variants['changed_exon'] = [changed_exon(i) for i in variants.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13617d88",
   "metadata": {},
   "outputs": [],
   "source": [
    "def HGVS_protein(ind):\n",
    "    one_to_three = {\n",
    "    'A': 'Ala', 'R': 'Arg', 'N': 'Asn', 'D': 'Asp', 'C': 'Cys',\n",
    "    'E': 'Glu', 'Q': 'Gln', 'G': 'Gly', 'H': 'His', 'I': 'Ile',\n",
    "    'L': 'Leu', 'K': 'Lys', 'M': 'Met', 'F': 'Phe', 'P': 'Pro',\n",
    "    'S': 'Ser', 'T': 'Thr', 'W': 'Trp', 'Y': 'Tyr', 'V': 'Val',\n",
    "    '*': 'Ter'}\n",
    "    ref1 = variants.iloc[ind]['ref_protein_seq']\n",
    "    alt1 = variants.iloc[ind]['protein_seq']\n",
    "    termin_prot = variants.iloc[ind]['termin_prot']\n",
    "    prot_change_pos = variants.iloc[ind]['prot_change_pos']\n",
    "    aa_next_pos = variants.iloc[ind]['aa_next_pos']\n",
    "    len_indel_prot = variants.iloc[ind]['len_indel_substit_prot']\n",
    "    ins_del_type = variants.iloc[ind]['kind']\n",
    "    \n",
    "    if prot_change_pos != 'no_change':\n",
    "        # len_indel = 0 -> \"probably frameshift\"\n",
    "        if len_indel_prot == 0:\n",
    "            ref_aa = one_to_three[ref1[prot_change_pos-1]]\n",
    "            alt_aa = one_to_three[alt1[prot_change_pos-1]]\n",
    "            if alt_aa == 'Ter':\n",
    "                prot_hgvs = f'NP_002991.2:p.{ref_aa}{prot_change_pos}{alt_aa}'\n",
    "            else:\n",
    "                prot_hgvs = f'NP_002991.2:p.{ref_aa}{prot_change_pos}{alt_aa}'\n",
    "                if aa_next_pos == 'probably frameshift':\n",
    "                    prot_hgvs += 'fs'\n",
    "                    if len(termin_prot) < len(ref1):\n",
    "                        len_to_termin = len(termin_prot) - prot_change_pos +1\n",
    "                        prot_hgvs += f'Ter{len_to_termin}'\n",
    "        # insertion or deletion with less/equal 3 aas inserted/deleted\n",
    "        elif len_indel_prot <= 3:\n",
    "            # no termination codon in inserted sequence\n",
    "            if ins_del_type == 'ins' and '*' not in alt1:\n",
    "                ref_aa_start = one_to_three[ref1[prot_change_pos-1]]\n",
    "                ref_aa_end = one_to_three[ref1[prot_change_pos]]\n",
    "                ins_aas = ''\n",
    "                for i in range(len_indel_prot):\n",
    "                    ins_aas += one_to_three[alt1[prot_change_pos-1+i]]\n",
    "                prot_hgvs = f'NP_002991.2:p.{ref_aa_start}{prot_change_pos}_{ref_aa_end}{prot_change_pos+1}ins{ins_aas}'\n",
    "            # termination codon in inserted sequence\n",
    "            elif ins_del_type == 'ins' and '*' in alt1:\n",
    "                ref_aa_start = one_to_three[ref1[prot_change_pos-1]]\n",
    "                ref_aa_end = one_to_three[ref1[prot_change_pos]]\n",
    "                ins_to_termin_len = len(termin_prot) - prot_change_pos\n",
    "                ins_aas = ''\n",
    "                if ins_to_termin_len == 0:\n",
    "                     prot_hgvs = f'NP_002991.2:p.{ref_aa_start}{prot_change_pos}_{ref_aa_end}{prot_change_pos+1}insTer'\n",
    "                else:\n",
    "                    for i in range(ins_to_termin_len):\n",
    "                        ins_aas += one_to_three[alt1[prot_change_pos-1+i]]\n",
    "                    prot_hgvs = f'NP_002991.2:p.{ref_aa_start}{prot_change_pos}_{ref_aa_end}{prot_change_pos+1}ins{ins_aas}'\n",
    "            # deleted aas\n",
    "            elif ins_del_type == 'del':\n",
    "                if len_indel_prot == 1:\n",
    "                    del_aa = one_to_three[ref1[prot_change_pos-1]]\n",
    "                    prot_hgvs = f'NP_002991.2:p.{del_aa}{prot_change_pos}del'\n",
    "                else:\n",
    "                    del_aa1 = one_to_three[ref1[prot_change_pos-1]]\n",
    "                    del_aa2 = one_to_three[ref1[prot_change_pos-1 + len_indel_prot-1]]\n",
    "                    prot_hgvs = f'NP_002991.2:p.{del_aa1}{prot_change_pos}_{del_aa2}{prot_change_pos+len_indel_prot-1}del'\n",
    "        # insertion or deletion with more than 3 aas inserted/deleted\n",
    "        else:\n",
    "            # no termination codon in inserted sequence\n",
    "            if ins_del_type == 'ins' and '*' not in alt1:\n",
    "                ref_aa_start = one_to_three[ref1[prot_change_pos-1]]\n",
    "                ref_aa_end = one_to_three[ref1[prot_change_pos]]\n",
    "                prot_hgvs = f'NP_002991.2:p.{ref_aa_start}{prot_change_pos}_{ref_aa_end}{prot_change_pos+1}insX[{len_indel_prot}]'\n",
    "            # termination codon in inserted sequence\n",
    "            elif ins_del_type == 'ins' and '*' in alt1:\n",
    "                ref_aa_start = one_to_three[ref1[prot_change_pos-1]]\n",
    "                ref_aa_end = one_to_three[ref1[prot_change_pos]]\n",
    "                ins_to_termin_len = len(termin_prot) - prot_change_pos\n",
    "                ins_aas = ''\n",
    "                if ins_so_termin_len == 0:\n",
    "                    prot_hgvs = f'NP_002991.2:p.{ref_aa_start}{prot_change_pos}_{ref_aa_end}{prot_change_pos+1}insTer'\n",
    "                elif ins_so_termin_len <= 3:\n",
    "                    for i in range(ins_to_termin_len):\n",
    "                        ins_aas += one_to_three[alt1[prot_change_pos-1+i]]\n",
    "                    prot_hgvs = f'NP_002991.2:p.{ref_aa_start}{prot_change_pos}_{ref_aa_end}{prot_change_pos+1}ins{ins_aas}'\n",
    "                else:\n",
    "                    prot_hgvs = f'NP_002991.2:p.{ref_aa_start}{prot_change_pos}_{ref_aa_end}{prot_change_pos+1}insTer{ins_to_termin_len}'\n",
    "                \n",
    "            # deleted aas\n",
    "            elif ins_del_type == 'del':\n",
    "                del_aa1 = one_to_three[ref1[prot_change_pos-1]]\n",
    "                del_aa2 = one_to_three[ref1[prot_change_pos-1 + len_indel_prot-1]]\n",
    "                prot_hgvs = f'NP_002991.2:p.{del_aa1}{prot_change_pos}_{del_aa2}{prot_change_pos+len_indel_prot-1}del'\n",
    "    else:\n",
    "        prot_hgvs = 'no_change'\n",
    "    return prot_hgvs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e446d69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein HGVS\n",
    "variants['protein_HGVS'] = [HGVS_protein(i) for i in variants.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46eb5223",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants.to_pickle('01_dataframes/'+prefix_variant_file+'_cDNA_protSeq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa50696",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants.to_excel('02_output_analysis/01_cDNA_table/'+prefix_variant_file+'_cDNA_protSeq.xlsx', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
