{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9481b88b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as pplt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "063e3210",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix_variant_file = 'transcripts_under_10_percent_expression'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf7328e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "imp_vars = pd.read_pickle('01_dataframes/'+prefix_variant_file+'_cDNA_protSeq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7610ac85",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_col(df, inp):\n",
    "    return [i for i in df.columns if inp.lower() in (i.lower())]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "981cb0a1",
   "metadata": {},
   "source": [
    "### Protein domain analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2713751d",
   "metadata": {},
   "outputs": [],
   "source": [
    "up_df = pd.read_pickle('01_dataframes/sdhb_uniprot_domains')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ca40a40",
   "metadata": {},
   "source": [
    "### NMD line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fadede57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GTF file with sequence information\n",
    "gtf_df = pd.read_pickle('01_dataframes/sdhb_gtf_pickle')\n",
    "gtf_df['exon_number'] = gtf_df['exon_number'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6974d0b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nmd_pos(gene):\n",
    "    gtf_spec = gtf_df[gtf_df['gene']==gene]\n",
    "    nmd_exon = max(gtf_spec['exon_number'].unique())-1\n",
    "    len_exon = (int(gtf_spec['end'][gtf_spec['exon_number']==nmd_exon])-\n",
    "                int(gtf_spec['start'][gtf_spec['exon_number']==nmd_exon]))\n",
    "    strand = gtf_spec['strand'].unique()[0]\n",
    "    len_cdna = 0\n",
    "\n",
    "    for i in gtf_spec['exon_number'].unique():\n",
    "        if i < nmd_exon:\n",
    "            len_cdna += (int(gtf_spec['end'][gtf_spec['exon_number']==i])-\n",
    "                         int(gtf_spec['start'][gtf_spec['exon_number']==i]))\n",
    "\n",
    "    if len_exon > 50:\n",
    "        if strand=='+':\n",
    "            nmd_pos = int(gtf_spec['end'][gtf_df['exon_number']==nmd_exon])-50\n",
    "            len_cdna += (nmd_pos-int(gtf_spec['start'][gtf_df['exon_number']==nmd_exon]))\n",
    "        else:\n",
    "            nmd_pos = int(gtf_spec['start'][gtf_df['exon_number']==nmd_exon])+50\n",
    "            len_cdna += (int(gtf_spec['end'][gtf_df['exon_number']==nmd_exon])-nmd_pos)\n",
    "\n",
    "    return (len_cdna, round(len_cdna/3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3733d09b",
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_lst = []\n",
    "for g in imp_vars['HUGO_Symbol'].unique():\n",
    "    c,p = nmd_pos(g)\n",
    "    nmd_lst.append([g,c,p])\n",
    "\n",
    "nmd_df = pd.DataFrame(nmd_lst, columns=['gene', 'NMD_cDNA', 'NMD_protein'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3678f20",
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_df.to_pickle('01_dataframes/'+prefix_variant_file+'_NMD')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3ec766f",
   "metadata": {},
   "outputs": [],
   "source": [
    "nmd_df = pd.read_pickle('01_dataframes/'+prefix_variant_file+'_NMD')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bdaece7",
   "metadata": {},
   "source": [
    "### Visualization Exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1547c086",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prot_domains(ind, df_var, up_df):\n",
    "    if ind in df_var.index:\n",
    "        prot = df_var.iloc[ind]['HUGO_Symbol']\n",
    "        ips_prot = up_df[up_df['HUGO_Symbol']==prot]\n",
    "        shortcut = df_var.iloc[ind]['shortcut']\n",
    "        \n",
    "        nmd_pos = int(nmd_df['NMD_protein'][nmd_df['gene']==prot])\n",
    "\n",
    "        prot_change_pos = df_var.iloc[ind]['prot_change_pos']\n",
    "        aa_next_pos = df_var.iloc[ind]['aa_next_pos']\n",
    "        termin_pos = df_var.iloc[ind]['termin_prot_len']\n",
    "\n",
    "        max_pos = df_var.iloc[ind]['ref_protein_seq_len']\n",
    "\n",
    "        seq_alt = df_var.iloc[ind]['protein_seq']\n",
    "        seq_ref = df_var.iloc[ind]['ref_protein_seq']\n",
    "    \n",
    "        if type(prot_change_pos)!='no_change':\n",
    "            try:\n",
    "                seq_alt = seq_alt[prot_change_pos-10:prot_change_pos+9]\n",
    "                seq_ref = seq_ref[prot_change_pos-10:prot_change_pos+9]\n",
    "            except:\n",
    "                seq_alt = seq_alt[prot_change_pos-10:]\n",
    "                seq_ref = seq_ref[prot_change_pos-10:]\n",
    "            seq_match = ''\n",
    "            for a,r in zip(seq_alt, seq_ref):\n",
    "                if a==r:\n",
    "                    seq_match += '|'\n",
    "                else:\n",
    "                    seq_match += ' '\n",
    "\n",
    "            inds = ips_prot.index\n",
    "            starts = ips_prot['start_pos']\n",
    "            ends = ips_prot['end_pos']\n",
    "            ids = ips_prot['info_domain']\n",
    "            ys = np.arange(0.01,1,1/len(ips_prot))\n",
    "\n",
    "            fig = plt.subplots(figsize=(14, 6))\n",
    "            ax1 = plt.subplot(1,1,1)\n",
    "\n",
    "            plt.ylim([0,1])\n",
    "            plt.xlim([0,max_pos])\n",
    "            ax1.get_yaxis().set_visible(False)\n",
    "            ax1.set_xlabel('amino acid position', fontsize=12)\n",
    "            ax1.xaxis.set_label_coords(0.5, -0.1)\n",
    "\n",
    "            ax1.set_title('Protein domains of %s (%s)' %(prot, shortcut), fontsize = 18, fontweight = 'bold', \n",
    "                          loc = 'center')\n",
    "\n",
    "            for i,s,e,y,t in zip(inds, starts, ends, ys, ids):\n",
    "                ax1.add_patch(pplt.Rectangle((s, y), e-s, 0.01, fc = 'black', ec = 'black', alpha = 1))\n",
    "                plt.text(max_pos+5, y, t, size = 13, ha='left')\n",
    "            if prot_change_pos != 'no_change':\n",
    "                ax1.add_patch(pplt.Rectangle((prot_change_pos, 0), 0.01, 1, fc = 'green', ec = 'green', alpha = 1))\n",
    "                plt.text(x=prot_change_pos, y=-0.10, s='aa substitution', ha='center', c='green')\n",
    "                if aa_next_pos not in ['probably frameshift', 'error']:\n",
    "                    len_rel_for_dom = aa_next_pos-prot_change_pos\n",
    "                    ax1.add_patch(pplt.Rectangle((prot_change_pos, 0), aa_next_pos-prot_change_pos, 1, \n",
    "                                                 fc = 'green', ec = 'green', alpha = 0.5))\n",
    "                    \n",
    "                elif aa_next_pos == 'probably frameshift':\n",
    "                    len_rel_for_dom = termin_pos-prot_change_pos\n",
    "                    ax1.add_patch(pplt.Rectangle((prot_change_pos, 0), termin_pos-prot_change_pos, 1, \n",
    "                                                 fc = 'magenta', ec = 'magenta', alpha = 0.5))\n",
    "                else:\n",
    "                    len_rel_for_dom = 1\n",
    "            if aa_next_pos not in ['probably frameshift', 'error']:\n",
    "                ax1.add_patch(pplt.Rectangle((termin_pos+(aa_next_pos-prot_change_pos), 0), 0.01, 1, fc = 'red', \n",
    "                                             ec = 'red', alpha = 1))\n",
    "                plt.text(x=termin_pos+(aa_next_pos-prot_change_pos), y=-0.07, s='stop', ha='center', c='red')\n",
    "            elif aa_next_pos == 'probably frameshift':\n",
    "                ax1.add_patch(pplt.Rectangle((termin_pos, 0), 0.01, 1, fc = 'red', ec = 'red', alpha = 1))\n",
    "                plt.text(x=termin_pos, y=-0.07, s='stop', ha='center', c='red')\n",
    "            \n",
    "            ax1.add_patch(pplt.Rectangle((nmd_pos, 0), 0.01, 1, fc = 'blue', ec = 'blue', alpha = 1))\n",
    "            plt.text(x=nmd_pos, y=-0.13, s='NMD', ha='center', c='blue')\n",
    "            plt.text(x=1, y=0.9, s=seq_alt, size = 15, ha='left', family='monospace')\n",
    "            plt.text(x=1, y=0.85, s=seq_match, size = 15, ha='left', family='monospace')\n",
    "            plt.text(x=1, y=0.8, s=seq_ref, size = 15, ha='left', family='monospace')\n",
    "            plt.gcf().subplots_adjust(bottom = 0.12, top = 0.94, left = 0.01, right = 0.6)\n",
    "            plt.savefig('02_output_analysis/02_domain_figs/dom_%s.png'%shortcut)\n",
    "            \n",
    "            out_domain_file = '02_output_analysis/03_domain_info/'+prefix_variant_file+'_'+shortcut+'_domain_info.tsv'\n",
    "            prot_doms_aff = []\n",
    "            for s,e,d in zip(starts, ends, ids):\n",
    "                if prot_change_pos >= s and prot_change_pos <= e:\n",
    "                    prot_doms_aff.append(d)\n",
    "                elif s >= prot_change_pos and e <= prot_change_pos+len_rel_for_dom:\n",
    "                    prot_doms_aff.append(d)\n",
    "            ips_prot[ips_prot['info_domain'].isin(prot_doms_aff)].to_csv(out_domain_file, sep='\\t', index=False)\n",
    "            \n",
    "        else:\n",
    "            print('no change in protein sequence')\n",
    "    else:\n",
    "        print('index not in dataframe')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8836aca",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(imp_vars)):\n",
    "    prot_domains(i, imp_vars, up_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47634b53",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "prot_domains(0, imp_vars, up_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
