{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from gtfparse import read_gtf\n",
    "import pysam as ps\n",
    "\n",
    "from sdhb_functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SDHB Reference Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gtf_df = read_gtf('Hsapiens/GRCh38/RefSeq/annotation_110/GRCh38_latest_genomic.gtf')\n",
    "# gtf_df[gtf_df['gene_id']=='SDHB'].to_pickle('01_New_reference_SDHB/sdhb_gtf_hg38.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_sdbh = pd.read_pickle('01_New_reference_SDHB/sdhb_gtf_hg38.gtf')\n",
    "ref_fa = ps.FastaFile('Hsapiens/GRCh38/RefSeq/assembly_GCF_000001405.40/GRCh38_latest_genomic.fna')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fasta reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plasmid sequence at intron 5\n",
    "s1 = 'TAAACTTAAGCTTGGTACCGAGCTCG'\n",
    "\n",
    "# intron 5: 202 bp\n",
    "s2 = ref_fa.fetch('NC_000001.11', 17027749-202-1, 17027749-1)\n",
    "\n",
    "# exon 4 complete, intron 4 complete, exon 5 complete\n",
    "s3 = ref_fa.fetch('NC_000001.11', 17027749-1, 17028736)\n",
    "\n",
    "# intron 3: - 325 bp\n",
    "s4 = ref_fa.fetch('NC_000001.11', 17028736, 17028736+325)\n",
    "\n",
    "# intron 3: + 324 bp\n",
    "s5 = ref_fa.fetch('NC_000001.11', 17033060-324-1, 17033060-1)\n",
    "\n",
    "# exon 3: complete\n",
    "s6 = ref_fa.fetch('NC_000001.11', 17033060-1, 17033145)\n",
    "\n",
    "# intron 2: - 325 bp\n",
    "s7 = ref_fa.fetch('NC_000001.11', 17033145, 17033145+325)\n",
    "\n",
    "# intron 2: + 325 bp\n",
    "s8 = ref_fa.fetch('NC_000001.11', 17044761-325-1, 17044761-1)\n",
    "\n",
    "# exon 2: complete\n",
    "s9 = ref_fa.fetch('NC_000001.11', 17044761-1, 17044888)\n",
    "\n",
    "# intron 1: 202 bp\n",
    "s10 = ref_fa.fetch('NC_000001.11', 17044888, 17044888+202)\n",
    "\n",
    "# plasmid sequence at intron 1:\n",
    "s11 = 'GTCTAGAGGGCCCGTTTAAACGCTAGCCAGCTT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s_e8 = ref_fa.fetch('NC_000001.11', 17018722-1, 17027749-202-1)\n",
    "\n",
    "s_i5_i2 = ref_fa.fetch('NC_000001.11', 17027749-202-1, 17044888+202)\n",
    "\n",
    "s_e1 = ref_fa.fetch('NC_000001.11', 17044888+202, 17054032)\n",
    "\n",
    "sdhb_ref = s_e8 + s1 + s_i5_i2 + s11 + s_e1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('01_New_reference_SDHB/sdhb_minigene_ref.fa', 'w') as o:\n",
    "    o.write('>chrSDHB_minigene\\n')\n",
    "    o.write(sdhb_ref.upper())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GTF annotation file SDHB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "altered_gtf_sdhb = gtf_sdbh.copy().reset_index(drop=True)\n",
    "altered_gtf_sdhb = altered_gtf_sdhb[:10]\n",
    "\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['seqname']=='NC_000001.11', 'seqname'] = 'chrSDHB_minigene'\n",
    "\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb.index==1, 'transcript_id'] = 'MG_unspliced'\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['feature']=='exon', 'transcript_id'] = 'MG_unspliced'\n",
    "\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['feature']=='gene', 'start'] = 1\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['feature']=='gene', 'end'] = 35370\n",
    "\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['exon_number']=='3', 'start'] = 8826\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['exon_number']=='3', 'end'] = 10367\n",
    "\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['exon_number']=='2', 'start'] = 14041\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['exon_number']=='2', 'end'] = 14776\n",
    "\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['exon_number']=='1', 'start'] = 25741\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['exon_number']=='1', 'end'] = 26429\n",
    "\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['feature']=='transcript', 'start'] = 8826\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['feature']=='transcript', 'end'] = 26429\n",
    "\n",
    "altered_gtf_sdhb = altered_gtf_sdhb.drop(index=[5,6,7,8,9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcr_row = altered_gtf_sdhb[altered_gtf_sdhb['feature']=='transcript']\n",
    "altered_gtf_sdhb = altered_gtf_sdhb.append(transcr_row, ignore_index=True)\n",
    "exon_row = altered_gtf_sdhb[altered_gtf_sdhb['exon_number']=='1']\n",
    "altered_gtf_sdhb = altered_gtf_sdhb.append([exon_row]*5, ignore_index=True)\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb.index.isin([5,6,7,8,9,10]), 'transcript_id'] = 'MG_spliced'\n",
    "for i in range(5):\n",
    "    altered_gtf_sdhb.loc[altered_gtf_sdhb.index==6+i, 'exon_number'] = str(1+i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='5')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'start'] = 8826\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='5')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'end'] = 9171\n",
    "\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='4')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'start'] = 9905\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='4')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'end'] = 10042\n",
    "\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='3')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'start'] = 14365\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='3')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'end'] = 14451\n",
    "\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='2')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'start'] = 26066\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='2')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'end'] = 26194\n",
    "\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='1')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'start'] = 26294\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['exon_number']=='1')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'end'] = 26429\n",
    "\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['feature']=='transcript')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'start'] = 8826\n",
    "altered_gtf_sdhb.loc[(altered_gtf_sdhb['feature']=='transcript')&(altered_gtf_sdhb['transcript_id']=='MG_spliced'),\n",
    "                     'end'] = 26429"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_attr = 'gene_id \"SDHB\"; gbkey \"Gene\"; gene \"SDHB\"; gene_biotype \"protein_coding\";'\n",
    "transcript_unspl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_unspliced\"; gbkey \"mRNA\"; gene \"SDHB\"; transcript_biotype \"SDHB\";'\n",
    "exon_1_unspl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_unspliced\"; gene \"SDHB\"; transcript_biotype \"SDHB\"; exon_number \"1\";'\n",
    "exon_2_unspl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_unspliced\"; gene \"SDHB\"; transcript_biotype \"SDHB\"; exon_number \"2\";'\n",
    "exon_3_unspl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_unspliced\"; gene \"SDHB\"; transcript_biotype \"SDHB\"; exon_number \"3\";'\n",
    "transcript_spl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_spliced\"; gbkey \"mRNA\"; gene \"SDHB\"; transcript_biotype \"SDHB\";'\n",
    "exon_1_spl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_spliced\"; gene \"SDHB\"; transcript_biotype \"SDHB\"; exon_number \"1\";'\n",
    "exon_2_spl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_spliced\"; gene \"SDHB\"; transcript_biotype \"SDHB\"; exon_number \"2\";'\n",
    "exon_3_spl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_spliced\"; gene \"SDHB\"; transcript_biotype \"SDHB\"; exon_number \"3\";'\n",
    "exon_4_spl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_spliced\"; gene \"SDHB\"; transcript_biotype \"SDHB\"; exon_number \"4\";'\n",
    "exon_5_spl_attr = 'gene_id \"SDHB\"; transcript_id \"MG_spliced\"; gene \"SDHB\"; transcript_biotype \"SDHB\"; exon_number \"5\";'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "attributes_list = [gene_attr, transcript_unspl_attr, exon_1_unspl_attr, exon_2_unspl_attr, exon_3_unspl_attr, \n",
    "                   transcript_spl_attr, exon_1_spl_attr, exon_2_spl_attr, exon_3_spl_attr, exon_4_spl_attr, \n",
    "                   exon_5_spl_attr]\n",
    "altered_gtf_sdhb['attributes'] = attributes_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare dataframe for gtf file format\n",
    "altered_gtf_sdhb['score'] = altered_gtf_sdhb['score'].replace(np.nan, '.')\n",
    "altered_gtf_sdhb.loc[altered_gtf_sdhb['source']=='', 'source'] = 'altered_ref_SDHB_minigene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "altered_gtf_sdhb[['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
    "                ].to_csv('01_New_reference_SDHB/altered_gtf_sdhb_mg_unspliced_spliced_together.gff', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "altered_gtf_sdhb[['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
    "                ].to_csv('01_New_reference_SDHB/altered_gtf_sdhb_mg_unspliced_spliced_together.gtf', sep='\\t', index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "altered_gtf_sdhb[altered_gtf_sdhb.index.isin([0,5,6,7,8,9,10])][['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
    "                ].to_csv('01_New_reference_SDHB/altered_gtf_sdhb_mg_only_spliced.gtf', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "altered_gtf_sdhb[altered_gtf_sdhb.index.isin([0,5,6,7,8,9,10])][['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
    "                ].to_csv('01_New_reference_SDHB/altered_gtf_sdhb_mg_only_spliced.gff', sep='\\t', index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
