{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from gtfparse import read_gtf\n",
    "import math\n",
    "import os\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as pplt\n",
    "from functools import reduce\n",
    "import pysam as ps\n",
    "\n",
    "from sdhb_functions_synthetic import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GTF file with sequence information\n",
    "# gtf_df = read_gtf('Hsapiens/GRCh38/RefSeq/annotation_110/GRCh38_latest_genomic.gtf')\n",
    "# gtf_df[gtf_df['gene_id']=='SDHB'].to_pickle('01_New_reference_SDHB/sdhb_gtf_hg38.gtf')\n",
    "gtf_sdbh = pd.read_pickle('01_New_reference_SDHB/sdhb_gtf_hg38.gtf')\n",
    "gtf_sdbh = gtf_sdbh[['feature','seqname','start','end','strand','frame','gene_id','transcript_id','exon_number']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Synthetic variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#synth_vars(sdhb_fa_path = '00_sdhb_seq_dfs/sdhb_fasta',\n",
    "#           sdhb_syn_out = '00_sdhb_seq_dfs/sdhb_syn.csv')\n",
    "sdhb_df = pd.read_csv('00_sdhb_seq_dfs/sdhb_syn.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variant in exonic or intronic region\n",
    "intr_ex_gtf_sdhb = exon_or_intron(sdhb_df = sdhb_df, gtf_sdbh = gtf_sdbh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exonic regions for VEP (also SpliceAI) -> 3045 bases\n",
    "#sdhb_df[sdhb_df['exon_intron'].str.startswith('exon')]['Identifier'].to_csv(\n",
    "    #'00_syn_tables_out/synth_sdhb_exon_variants.txt', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# include certain region into intron\n",
    "#intron_df = intron_df_vep(1400,1300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VEP output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adjust VEP output file\n",
    "\n",
    "#ex_sdhb_vep = vep_output('00_Andres_SDHB/exonic_sdhb_bep.txt')\n",
    "#in_sdhb_vep = vep_output('00_Andres_SDHB/intronic_sdhb_vep.txt')\n",
    "\n",
    "#pd.DataFrame({'files':[file for file in os.listdir('00_Andres_SDHB/02_vep_results/') \n",
    "#                       if file.endswith('.txt')]}).to_csv('00_Andres_SDHB/02_vep_results/vep_dict_list', index=False)\n",
    "\n",
    "# for all files in 02_vep_results -> run vep_output()\n",
    "vep_dict = run_all_vep()\n",
    "all_vars = pd.DataFrame()\n",
    "for key in vep_dict.keys():\n",
    "    all_vars = all_vars.append(vep_dict[key])\n",
    "all_vars = all_vars.sort_values('POS').reset_index(drop=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
